MdeModulePkg/Core/Dxe/Misc/MemoryProtectionSupport.c

```diff
@@ -0,0 +1,2853 @@
+      (DescriptorSize == NULL) || (Bitmap == NULL))
+  {
+    return EFI_INVALID_PARAMETER;
+  }
+
+  MemoryMapEntry = MemoryMap;
+  MemoryMapEnd   = (EFI_MEMORY_DESCRIPTOR *)((UINT8 *)MemoryMap + *MemoryMapSize);
+
+  while (MemoryMapEntry < MemoryMapEnd) {
+    // If attributes are nonzero or the virtual address (always zero during boot services according to UEFI Spec 2.9)
+    // is set to the pattern indicating it is a special memory region, set the corresponding bit in the bitmap
+    if ((MemoryMapEntry->Attribute != 0) || (MemoryMapEntry->VirtualStart == SPECIAL_REGION_PATTERN)) {
+      SET_BITMAP_INDEX (Bitmap, Index);
+    }
+
+    Index++;
+    MemoryMapEntry = NEXT_MEMORY_DESCRIPTOR (MemoryMapEntry, *DescriptorSize);
+  }
+
+  return EFI_SUCCESS;
+}
+
+/**
+  Set access attributes in the memory map based on the memory protection policy and
+  mark visited regions in the bitmap.
+
+  @param[in]        MemoryMapSize           A pointer to the size, in bytes, of the MemoryMap buffer
+  @param[in, out]   MemoryMap               A pointer to the current memory map
+  @param[in]        DescriptorSize          Size, in bytes, of an individual EFI_MEMORY_DESCRIPTOR
+  @param[in, out]   Bitmap                  Pointer to the beginning of the bitmap to be updated
+
+  @retval           EFI_SUCCESS             Access attributes and bitmap updated
+  @retval           EFI_INVALID_PARAMETER   An input was NULL
+**/
+STATIC
+EFI_STATUS
+SetAccessAttributesInMemoryMap (
+  IN CONST  UINTN                  *MemoryMapSize,
+  IN OUT    EFI_MEMORY_DESCRIPTOR  *MemoryMap,
+  IN CONST  UINTN                  *DescriptorSize,
+  IN OUT    UINT8                  *Bitmap
+  )
+{
+  EFI_MEMORY_DESCRIPTOR  *MemoryMapEntry;
+  EFI_MEMORY_DESCRIPTOR  *MemoryMapEnd;
+  UINTN                  Index = 0;
+
+  if ((MemoryMapSize == NULL) || (MemoryMap == NULL) ||
+      (DescriptorSize == NULL) || (Bitmap == NULL))
+  {
+    return EFI_INVALID_PARAMETER;
+  }
+
+  MemoryMapEntry = MemoryMap;
+  MemoryMapEnd   = (EFI_MEMORY_DESCRIPTOR *)((UINT8 *)MemoryMap + *MemoryMapSize);
+
+  while (MemoryMapEntry < MemoryMapEnd) {
+    if (!IS_BITMAP_INDEX_SET (Bitmap, Index)) {
+      MemoryMapEntry->Attribute = GetPermissionAttributeForMemoryType (MemoryMapEntry->Type);
+      SET_BITMAP_INDEX (Bitmap, Index);
+    }
+
+    Index++;
+    MemoryMapEntry = NEXT_MEMORY_DESCRIPTOR (MemoryMapEntry, *DescriptorSize);
+  }
+
+  return EFI_SUCCESS;
+}
+
+/**
+  Merge contiguous memory map entries with the same attributes.
+
+  @param  MemoryMap              A pointer to the memory map
+  @param  MemoryMapSize          A pointer to the size, in bytes, of the
+                                 MemoryMap buffer. On input, this is the size of the current
+                                 memory map.  On output, it is the size of new memory map after merge.
+  @param  DescriptorSize         Size, in bytes, of an individual EFI_MEMORY_DESCRIPTOR.

```

Comment:
Most functions have them, so include direction here too.

```suggestion
  @param[in,out]  MemoryMap      A pointer to the memory map
  @param[in,out]  MemoryMapSize  A pointer to the size, in bytes, of the
                                 MemoryMap buffer. On input, this is the size of the current
                                 memory map.  On output, it is the size of new memory map after merge.
  @param[in]  DescriptorSize     Size, in bytes, of an individual EFI_MEMORY_DESCRIPTOR.

```